Generator/Sources/NeedleFramework/Generating/Serializers/OutputSerializer.swift

```diff
@@ -37,23 +37,74 @@ class OutputSerializer: Serializer {
     ///
     /// - returns: The source code `String`.
     func serialize() -> String {
-        let registrationBody = providers
-            .map { (provider: SerializedProvider) in
-                provider.registration
-            }
-            .joined()
-            .replacingOccurrences(of: "\n", with: "\n    ")
+        // Dependency factories have deterministic names, so we can only define a
+        // factory once, or the code won't compile.
+        var registeredFactories: Set<String> = []
+        // We generate parent traversal functions to improve compile time on large
+        // codebases. This tells us how many levels deep the tree goes.
+        var maxLevel: Int = 1
 
         let providersSection = providers
             .map { (provider: SerializedProvider) in
-                provider.content
+                if let providerMaxLevel = provider.attributes.maxLevel {
+                    if providerMaxLevel > maxLevel {
+                        maxLevel = providerMaxLevel
+                    }
+                }
+                if let factoryName = provider.attributes.factoryName {
+                    if registeredFactories.contains(factoryName) {
+                        return ""
```

Comment:
This collision detection may no longer needed if we use the path to name the factories, but need to look closely and make sure.